### ISSUE TITLE
`zrep refresh` is buggy

### ISSUE THREAD

**Issue by wxcafe** [bolthole/zrep#155]
So I've been using zrep for about two weeks now, and I regularly have problems with `zrep refresh`.

Some of these I assume aren't zrep's fault (i.e. I sometimes get `cannot receive incremental stream: incompatible embedded data stream feature with encrypted receive.` outputs, which seem related to openzfs/zfs#6366), but once I have an error in a refresh run it will not complete again by itself, i.e. every subsequent run errors out until I fix them manually.

I get errors of the type 

```
cannot receive incremental stream: most recent snapshot of pool/dataset does not
match incremental source
Error: Unforseen error pulling snapshot remotepool/remotedataset@zrep-kirby_000052 from remoteserver
```

Both of these look a bit like #149. From what I understand, the first error means the snapshot isn't synced, but `zrep refresh` still marks it as successfully sent, and subsequent runs try to send an incremental snapshot based on that one (which isn't present on the local side, and so predictably fails). It seems like `_refreshpull()` should call the same `zfs rename _unsent` function that `sync()` does if the send fails, and not set `$senttimeprop` in that case

**Comment by wxcafe** [comment]
opened #156 to address part of this. The problem though is that the error happens on the receiving side, the sending side probably sees nothing wrong (since the `send` happens fine, it's the `recv` that fails). The receiving side probably needs to ssh back to the sending side to rename the snap as `_unsent` and remove `$senttimeprop` if receiving failed, that's a new function

Though maybe it'd be easier to have `zrep_refresh()` call `_refreshpull` with the latest zrep snapshot on the receiving side as an argument, so that `_refreshpull` sends the diff with that version?

**Comment by wxcafe** [comment]
opened #157 to try and implement that second part. This is a rough draft that definitely needs some testing and probably lots of refinements, I just don't have the time to test it right now

**Comment by wxcafe** [comment]
fixed by a19b0bf42af1035250a5283777aa7bfc3c10315f / a5e5d2013ab4d88ef53e9172424eb91fc359523a / e2e7af9718d5add3de554d21918c5942412bb755 and 8496e27f7b2217153cb5a1714abaa5118938642a
